### PR TITLE
index metadata loading for MySQL dialect(#9617)

### PR DESCRIPTION
**Pull request by wordKuaiLei**
For #9617

### CODE REVIEW

**Comment by wordKuaiLei** [comment]
index metadata loading for MySQL dialect(#9617)

**Comment by wordKuaiLei** [comment]
index metadata loading for MySQL dialect(#9617)

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java [hunk 1]
```diff
@@ -54,11 +57,29 @@
         Map<String, TableMetaData> result = new LinkedHashMap<>();
         for (Entry<String, Collection<ColumnMetaData>> entry : loadColumnMetaDataMap(dataSource, existedTables).entrySet()) {
             // TODO load index
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java:59]: Please remove todo if the feature is complete

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java [hunk 2]
```diff
@@ -54,11 +57,29 @@
...
-            result.put(entry.getKey(), new TableMetaData(entry.getValue(), Collections.emptyList()));
+            Collection<IndexMetaData> indexMetaData = loadIndexMetaData(dataSource, entry.getKey());
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java:60]: Please init indexMetaData by constructor

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java [hunk 3]
```diff
@@ -54,11 +57,29 @@
...
+            result.put(entry.getKey(), new TableMetaData(entry.getValue(), indexMetaData));
         }
         return result;
     }
-    
+
+    private Collection<IndexMetaData> loadIndexMetaData(final DataSource dataSource, final String tableName) throws SQLException {
+        Collection<IndexMetaData> result = Lists.newArrayList();
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java:67]: Please use LinkedList if do not need to get by random

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java [hunk 4]
```diff
@@ -54,11 +57,29 @@
...
+        try (
+                Connection connection = dataSource.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(BASIC_INDEX_META_DATA_SQL)) {
+            preparedStatement.setString(1, tableName);
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                while (resultSet.next()) {
+                    String indexName = resultSet.getString("INDEX_NAME");
+                    IndexMetaData indexMetaData = new IndexMetaData(indexName);
+                    result.add(indexMetaData);
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoader.java:76]: It is better to inline `new IndexMetaData` and `result.add`

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java [hunk 5]
```diff
@@ -43,6 +43,9 @@
     
     @Getter(AccessLevel.NONE)
     private final List<String> columnNames = new ArrayList<>();
+
+    @Getter(AccessLevel.NONE)
+    private final List<String> indexNames = new ArrayList<>();
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java:48]: Please do not change `TableMetaData`, we just load, the usage is not need to change.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java [hunk 6]
```diff
@@ -71,7 +74,9 @@ public TableMetaData(final Collection<ColumnMetaData> columnMetaDataList, final
     private Map<String, IndexMetaData> getIndexes(final Collection<IndexMetaData> indexMetaDataList) {
         Map<String, IndexMetaData> result = new LinkedHashMap<>(indexMetaDataList.size(), 1);
         for (IndexMetaData each : indexMetaDataList) {
-            result.put(each.getName().toLowerCase(), each);
+            String lowerIndexName = each.getName().toLowerCase();
+            indexNames.add(lowerIndexName);
+            result.put(lowerIndexName, each);
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java:79]: Please do not change `TableMetaData`, we just load, the usage is not need to change.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java [hunk 7]
```diff
@@ -85,6 +90,16 @@ public TableMetaData(final Collection<ColumnMetaData> columnMetaDataList, final
...
+
+    /**
+     * Get index meta data.
+     *
+     * @param indexIndex  index index
+     * @return index meta data
+     */
+    public IndexMetaData getIndexMetaData(final int indexIndex) {
+        return indexes.get(indexNames.get(indexIndex));
+    }
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/model/TableMetaData.java:102]: Please do not change `TableMetaData`, we just load, the usage is not need to change.

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoaderTest.java [hunk 8]
```diff
@@ -103,5 +117,7 @@ private void assertTableMetaDataMap(final Map<String, TableMetaData> actual) {
         assertThat(actual.get("tbl").getColumns().size(), is(2));
         assertThat(actual.get("tbl").getColumnMetaData(0), is(new ColumnMetaData("id", 4, true, true, true)));
         assertThat(actual.get("tbl").getColumnMetaData(1), is(new ColumnMetaData("name", 12, false, false, false)));
+        assertThat(actual.get("tbl").getIndexes().size(), is(1));
+        assertThat(actual.get("tbl").getIndexMetaData(0), is(new IndexMetaData("id")));
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoaderTest.java:121]: Please fix assertion by original `getIndexes`

File: shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoaderTest.java [hunk 9]
```diff
@@ -88,6 +95,13 @@ private ResultSet mockTableMetaDataResultSet() throws SQLException {
         when(result.getString("COLLATION_NAME")).thenReturn("utf8_general_ci", "utf8");
         return result;
     }
+
+    private ResultSet mockIndexMetaDataResultSet() throws SQLException {
+        ResultSet result = mock(ResultSet.class);
+        when(result.next()).thenReturn(true, false, false);
```
Comment (terrymanu) [inline shardingsphere-infra/shardingsphere-infra-common/src/test/java/org/apache/shardingsphere/infra/metadata/schema/builder/loader/dialect/MySQLTableMetaDataLoaderTest.java:101]: `true, false, false` should be `true, false`

**Review by terrymanu — CHANGES_REQUESTED** [review]

**Review by terrymanu — APPROVED** [review]

**Pull request merged**